include/smack/smack.h

```diff
@@ -135,32 +224,148 @@ void __SMACK_decls() {
+  D("function {:inline} $zext.i16i32(p: i16) returns (i32) {(0bv16)++p}");
+  D("function {:inline} $zext.i32i64(p: i32) returns (i64) {p}");
+  // Sext
+  D("function {:inline} $sext.i8i64(p: i8) returns (i64) {if $sge.i8(p, 0bv8) then $zext.i8i64(p) else (($neg.i32(1bv32))[32:8])++p}");
+  D("function {:inline} $sext.i8i32(p: i8) returns (i32) {if $sge.i8(p, 0bv8) then $zext.i8i32(p) else (($neg.i32(1bv32))[32:8])++p}");
+  D("function {:inline} $sext.i8i16(p: i8) returns (i16) {if $sge.i8(p, 0bv8) then $zext.i8i16(p) else $neg.i8(1bv8)++p}");
+  D("function {:inline} $sext.i16i64(p: i16) returns (i64) {if $sge.i16(p, 0bv16) then $zext.i16i64(p) else $neg.i16(1bv16)++p}");
+  D("function {:inline} $sext.i16i32(p: i16) returns (i32) {if $sge.i16(p, 0bv16) then $zext.i16i32(p) else $neg.i16(1bv16)++p}");
+  D("function {:inline} $sext.i32i64(p: i32) returns (i64) {p}");
+
+  D("function {:inline} $p2i(p: ref) returns (i64) {p}");
+  D("function {:inline} $i2p(p: i64) returns (ref) {p}");
+  D("function {:inline} $p2b(p: ref) returns (bool) {p != 0bv32}");
+#else
+  D("function $base(int) returns (int);");
   D("const unique $NULL: int;");
   D("axiom $NULL == 0;");
-  D("const $UNDEF: int;");
-
   D("function {:inline} $pa(pointer: int, index: int, size: int) returns (int) {pointer + index * size}");
-  D("function {:inline} $trunc(p: int, size: int) returns (int) {p}");
+  D("function {:inline} $b2p(b: bool) returns (int) {if b then 1 else 0}");
   D("function {:inline} $p2i(p: int) returns (int) {p}");
   D("function {:inline} $i2p(p: int) returns (int) {p}");
   D("function {:inline} $p2b(p: int) returns (bool) {p != 0}");
-  D("function {:inline} $b2p(b: bool) returns (int) {if b then 1 else 0}");
+#endif
+  D("function {:inline} $trunc(p: int) returns (int) {p}");
+  D("function {:inline} $zext(p: int) returns (int) {p}");
+  D("function {:inline} $sext(p: int) returns (int) {p}");
 
   // Memory debugging symbols
   D("type $mop;");
   D("procedure boogie_si_record_mop(m: $mop);");
   D("procedure boogie_si_record_bool(b: bool);");
+#ifdef BITVECTOR
+  D("procedure boogie_si_record_i8(i: i8);");
+  D("procedure boogie_si_record_i16(i: i16);");
+  D("procedure boogie_si_record_i32(i: i32);");
+  D("procedure boogie_si_record_i64(i: i64);");
+  D("procedure boogie_si_record_ref(i: ref);");
+#else
   D("procedure boogie_si_record_int(i: int);");
+#endif
   D("procedure boogie_si_record_float(f: float);");
   D("const $MOP: $mop;");
-  
+
+#ifdef BITVECTOR
+  D("const $GLOBALS_BOTTOM: ref;");
+  D("function {:inline} $isExternal(p: ref) returns (bool) { $slt.i32(p, $sub.i32($GLOBALS_BOTTOM, 32768bv32)) }");
+#else
   D("const $GLOBALS_BOTTOM: int;");
   D("function {:inline} $isExternal(p: int) returns (bool) { p < $GLOBALS_BOTTOM - 32768 }");
-  
+#endif
+
 #if MEMORY_MODEL_NO_REUSE_IMPLS
+#ifdef BITVECTOR
+  D("var $Alloc: [ref] bool;");
+  D("var $CurrAddr:ref;");
+
+  D("procedure $malloc(n: i32) returns (p: ref)\n"
```

Comment:
Looking at the use of the `$add` functions inside of `$malloc` and friends also suggests that we should be using `$pa` rather than `$add` to compute pointer arithmetic, e.g., between `ref`-type addresses and `size_t` offsets.